interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,143 @@
+# file
+================
+
+## Overview
+
+The `file` module provides a platform-independent API for asynchronous file operations.
+
+-`file_create`: returns a file handle for the given file name.
+-`file_destroy`: closes the given file handle.
+-`file_write_async`: enqueues an asynchronous write request for a file at a given position.
+-`file_read_async`: enqueues an asynchronous read request for a file at a given position and size.
+-`file_extend_filesize`: expands the given file to be of desired size.
+
+## Exposed API
+
+```c
+#define FILE_WRITE_ASYNC_VALUES \
+    FILE_WRITE_ASYNC_INVALID_ARGS, \
+    FILE_WRITE_ASYNC_WRITE_ERROR, \
+    FILE_WRITE_ASYNC_ERROR,\
+    FILE_WRITE_ASYNC_OK
+MU_DEFINE_ENUM(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES);
+
+#define FILE_READ_ASYNC_VALUES \
+    FILE_READ_ASYNC_INVALID_ARGS, \
+    FILE_READ_ASYNC_READ_ERROR, \
+    FILE_READ_ASYNC_ERROR,\
+    FILE_READ_ASYNC_OK
+MU_DEFINE_ENUM(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES);
+
+typedef struct FILE_HANDLE_DATA_TAG* FILE_HANDLE;
+typedef void(*FILE_REPORT_FAULT)(void* user_report_fault_context, const char* information);
+
+typedef void(*FILE_WRITE_CB)(void* user_context, bool is_successful);
+typedef void(*FILE_READ_CB)(void* user_context, bool is_successful, CONSTBUFFER_HANDLE content);
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);

```

Comment:
>file_extend_filesize [](start = 38, length = 20)

I see this as a separate API from the read/write.

Reply:
FYI, I suggested this because it will act on the same internal windows HANDLE (and unknown for Linux) so if it's in a different module then this would need a "get handle" function, which seems to sort of expose things that shouldn't be exposed. Unless you have a different idea for organizing that.

---
In reply to: [450457028](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450457028) [](ancestors = 450457028)